Use proper HirId for async track_caller attribute check

nbdd0121 opened the pull request:
Fix #105134

nbdd0121 committed:
Use proper HirId for async track_caller attribute check

nbdd0121 committed:
Fix async track caller for assoc fn and trait impl fn

compiler-errors commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -1012,6 +1013,7 @@ impl<'hir> LoweringContext<'_, 'hir> {
 
                 let async_body = this.make_async_expr(
```
Async closures should probably also support `#[track_caller]`

nbdd0121 replied:
It currently doesn't, so that's why I haven't changed it in this PR.

It isn't difficult change to make, but it does require lifting the attribute lower up from after `ExprKind` lowering to before it. Do you happen to know why we currently lower ID and attributes after `ExprKind`?

nbdd0121 replied:
Added. I then changed `Option<HirId>` to `HirId` to simplify code. We don't need to special case async blocks as `#[track_caller]` is not legal on async block. But I added a test anyway to avoid accidentally allow this on normal async block.

compiler-errors replied:
Given the incremental issues, maybe we can revert the last change and go with `Option<HirId>`? Or are those test failures unrelated to the last change?

compiler-errors commented:
r? @compiler-errors @bors delegate+

Thanks, r=me with or without my comment.

cc @bryangarza, you will need to rebase your PR.

bors commented:
:v: @nbdd0121 can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"compiler-errors","delegate":"nbdd0121"} -->

compiler-errors commented:
I'll review the changes in a bit

nbdd0121 commented:
Not sure why making this change cause incr-comp test to fail. Maybe different HirId ordering causes the typeck result happens to match? Might need some incr-comp expert to take a look @rust-lang/wg-incr-comp

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3603894841/jobs/6072636267) [(plain)](https://github.com/rust-lang/rust/commit/11fa1c36f3927496f419396db71e03db0f96792e/checks/9852853307/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 finished in 0.724 seconds
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 158 tests
............................................F............F.F............................ 88/158
Some tests failed in compiletest suite=incremental mode=incremental host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
failures:

---- [incremental] src/test/incremental/hashes/loop_expressions.rs stdout ----
---- [incremental] src/test/incremental/hashes/loop_expressions.rs stdout ----

error in revision `cfail2`: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/incremental/hashes/loop_expressions.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "cfail2" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/loop_expressions/loop_expressions.inc" "-Z" "incremental-verify-ich" "-O" "--error-format" "json" "--json" "future-incompat" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/loop_expressions" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/loop_expressions/auxiliary" "-Z" "query-dep-graph" "-O" "-Zincremental-ignore-spans"
stdout: none
--- stderr -------------------------------
error: `typeck(change_continue_label)` should be dirty but is not
   |
LL | pub fn change_continue_label() {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


error: aborting due to previous error
------------------------------------------


---- [incremental] src/test/incremental/hashes/while_let_loops.rs stdout ----

error in revision `cfail2`: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/incremental/hashes/while_let_loops.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "cfail2" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/while_let_loops/while_let_loops.inc" "-Z" "incremental-verify-ich" "-O" "--error-format" "json" "--json" "future-incompat" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/while_let_loops" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/while_let_loops/auxiliary" "-Z" "query-dep-graph" "-O" "-Zincremental-ignore-spans"
stdout: none
--- stderr -------------------------------
error: `typeck(change_break_label)` should be dirty but is not
   |
   |
LL | pub fn change_break_label() {


error: `typeck(change_continue_label)` should be dirty but is not
   |
LL | pub fn change_continue_label() {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


error: aborting due to 2 previous errors
------------------------------------------


---- [incremental] src/test/incremental/hashes/while_loops.rs stdout ----

error in revision `cfail2`: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/incremental/hashes/while_loops.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "cfail2" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/while_loops/while_loops.inc" "-Z" "incremental-verify-ich" "-O" "--error-format" "json" "--json" "future-incompat" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/while_loops" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/while_loops/auxiliary" "-Z" "query-dep-graph" "-O" "-Zincremental-ignore-spans"
stdout: none
--- stderr -------------------------------
error: `typeck(change_break_label)` should be dirty but is not
   |
   |
LL | pub fn change_break_label() {


error: `typeck(change_continue_label)` should be dirty but is not
   |
LL | pub fn change_continue_label() {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

</details>

cjgillot commented:
You can just bless the incremental tests. The change is in the right direction: stuff that were dirty are now clean.
In the places there is a reported error, remove the reported query from the `rustc_clean` attribute's `except` list.

nbdd0121 commented:
Thanks, @cjgillot! I thought that they were there to ensure that the results are dirty. So these are more like status-quo checks instead to catch accidental changes?

cjgillot commented:
Pretty much, yes.

compiler-errors commented:
Given @cjgillot's comment and the blessed tests, this seems good to go now. Thanks for fixing this @nbdd0121.

@bors r+

bors commented:
:pushpin: Commit 76ee06b330ab37588cc380a61f05ef5a56ea4d3c has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 76ee06b330ab37588cc380a61f05ef5a56ea4d3c -->
<!-- homu: {"type":"Approved","sha":"76ee06b330ab37588cc380a61f05ef5a56ea4d3c","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 76ee06b330ab37588cc380a61f05ef5a56ea4d3c with merge a3c1e064379c6c903f1d52530d1847e1bb130e50...
<!-- homu: {"type":"BuildStarted","head_sha":"76ee06b330ab37588cc380a61f05ef5a56ea4d3c","merge_sha":"a3c1e064379c6c903f1d52530d1847e1bb130e50"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3616371191/jobs/6094442851)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3616371191/jobs/6094442851","builder_name":"checks-actions"} -->

nbdd0121 committed:
Add fixme note

nbdd0121 commented:
Given that #105134 is a critical regression and thus we want it to be fixed soon, I think it's best to split the async closure support (with HIR attribute lowering changes) into a separate PR. I have added a FIXME note and reverted the PR to the pre `Option<HirId>` -> `HirId` change state.

@bors r=compiler-errors

bors commented:
:pushpin: Commit 34c377335bc6b855722eb1f4a0866de29af0f2c9 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 34c377335bc6b855722eb1f4a0866de29af0f2c9 -->
<!-- homu: {"type":"Approved","sha":"34c377335bc6b855722eb1f4a0866de29af0f2c9","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3616371191/jobs/6094262202) [(plain)](https://github.com/rust-lang-ci/rust/commit/a3c1e064379c6c903f1d52530d1847e1bb130e50/checks/9879096279/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- compile_test stdout ----
diff of stdout:

 if let Some(higher::ForLoop { pat: pat, arg: arg, body: body, .. }) = higher::ForLoop::hir(expr)
     && let PatKind::Binding(BindingAnnotation::NONE, _, name, None) = pat.kind
     && name.as_str() == "y"
     && let ExprKind::Struct(qpath, fields, None) = arg.kind
     && matches!(qpath, QPath::LangItem(LangItem::Range, _))
     && fields.len() == 2
     && fields[0].ident.as_str() == "start"
     && let ExprKind::Lit(ref lit) = fields[0].expr.kind
     && let LitKind::Int(0, LitIntType::Unsuffixed) = lit.node
     && fields[1].ident.as_str() == "end"
     && let ExprKind::Lit(ref lit1) = fields[1].expr.kind
     && let LitKind::Int(10, LitIntType::Unsuffixed) = lit1.node
     && let ExprKind::Block(block, None) = body.kind
     && block.stmts.len() == 1
     && let StmtKind::Local(local) = block.stmts[0].kind
     && let Some(init) = local.init
error: test failed, to rerun pass `--test compile-test`
     && let ExprKind::Path(ref qpath1) = init.kind
     && match_qpath(qpath1, &["y"])
     && let PatKind::Binding(BindingAnnotation::NONE, _, name1, None) = local.pat.kind
     && name1.as_str() == "z"
     && block.expr.is_none()
     // report your lint here
 }
 }
+if let ExprKind::Loop(body, None, LoopSource::ForLoop, _) = expr.kind
+    && body.stmts.len() == 1
+    && let StmtKind::Expr(e) = body.stmts[0].kind
+    && let ExprKind::Match(scrutinee, arms, MatchSource::ForLoopDesugar) = e.kind
+    && let ExprKind::Call(func, args) = scrutinee.kind
+    && let ExprKind::Path(ref qpath) = func.kind
+    && matches!(qpath, QPath::LangItem(LangItem::IteratorNext, _))
+    && args.len() == 1
+    && let ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mut, inner) = args[0].kind
+    && let ExprKind::Path(ref qpath1) = inner.kind
+    && match_qpath(qpath1, &["iter"])
+    && arms.len() == 2
+    && let PatKind::Struct(ref qpath2, fields, false) = arms[0].pat.kind
+    && matches!(qpath2, QPath::LangItem(LangItem::OptionNone, _))
+    && fields.is_empty()
+    && arms[0].guard.is_none()
+    && let ExprKind::Break(destination, None) = arms[0].body.kind
+    && destination.label.is_none()
+    && let PatKind::Struct(ref qpath3, fields1, false) = arms[1].pat.kind
+    && matches!(qpath3, QPath::LangItem(LangItem::OptionSome, _))
+    && fields1.len() == 1
+    && fields1[0].ident.as_str() == "0"
+    && let PatKind::Binding(BindingAnnotation::NONE, _, name, None) = fields1[0].pat.kind
+    && name.as_str() == "y"
+    && arms[1].guard.is_none()
+    && let ExprKind::Block(block, None) = arms[1].body.kind
+    && block.stmts.len() == 1
+    && let StmtKind::Local(local) = block.stmts[0].kind
+    && let Some(init) = local.init
+    && let ExprKind::Path(ref qpath4) = init.kind
+    && match_qpath(qpath4, &["y"])
+    && let PatKind::Binding(BindingAnnotation::NONE, _, name1, None) = local.pat.kind
+    && name1.as_str() == "z"
+    && block.expr.is_none()
+    && body.expr.is_none()
+    // report your lint here
+}
+}
 if let Some(higher::ForLoop { pat: pat, arg: arg, body: body, .. }) = higher::ForLoop::hir(expr)
     && let PatKind::Wild = pat.kind
     && let ExprKind::Struct(qpath, fields, None) = arg.kind
     && matches!(qpath, QPath::LangItem(LangItem::Range, _))
     && fields.len() == 2
     && fields[0].ident.as_str() == "start"
     && let ExprKind::Lit(ref lit) = fields[0].expr.kind
     && let LitKind::Int(0, LitIntType::Unsuffixed) = lit.node
     && fields[1].ident.as_str() == "end"
     && let ExprKind::Lit(ref lit1) = fields[1].expr.kind
     && let LitKind::Int(10, LitIntType::Unsuffixed) = lit1.node
     && let ExprKind::Block(block, None) = body.kind
     && block.stmts.len() == 1
     && let StmtKind::Semi(e) = block.stmts[0].kind
     && let ExprKind::Break(destination, None) = e.kind
     && destination.label.is_none()
     && block.expr.is_none()
     // report your lint here
 }
 }
+if let ExprKind::Loop(body, None, LoopSource::ForLoop, _) = expr.kind
+    && body.stmts.len() == 1
+    && let StmtKind::Expr(e) = body.stmts[0].kind
+    && let ExprKind::Match(scrutinee, arms, MatchSource::ForLoopDesugar) = e.kind
+    && let ExprKind::Call(func, args) = scrutinee.kind
+    && let ExprKind::Path(ref qpath) = func.kind
+    && matches!(qpath, QPath::LangItem(LangItem::IteratorNext, _))
+    && args.len() == 1
+    && let ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mut, inner) = args[0].kind
+    && let ExprKind::Path(ref qpath1) = inner.kind
+    && match_qpath(qpath1, &["iter"])
+    && arms.len() == 2
+    && let PatKind::Struct(ref qpath2, fields, false) = arms[0].pat.kind
+    && matches!(qpath2, QPath::LangItem(LangItem::OptionNone, _))
+    && fields.is_empty()
+    && arms[0].guard.is_none()
+    && let ExprKind::Break(destination, None) = arms[0].body.kind
+    && destination.label.is_none()
+    && let PatKind::Struct(ref qpath3, fields1, false) = arms[1].pat.kind
+    && matches!(qpath3, QPath::LangItem(LangItem::OptionSome, _))
+    && fields1.len() == 1
+    && fields1[0].ident.as_str() == "0"
+    && let PatKind::Wild = fields1[0].pat.kind
+    && arms[1].guard.is_none()
+    && let ExprKind::Block(block, None) = arms[1].body.kind
+    && block.stmts.len() == 1
+    && let StmtKind::Semi(e1) = block.stmts[0].kind
+    && let ExprKind::Break(destination1, None) = e1.kind
+    && destination1.label.is_none()
+    && block.expr.is_none()
+    && body.expr.is_none()
+    // report your lint here
+}
+}
 if let Some(higher::ForLoop { pat: pat, arg: arg, body: body, .. }) = higher::ForLoop::hir(expr)
     && let PatKind::Wild = pat.kind
     && let ExprKind::Struct(qpath, fields, None) = arg.kind
     && matches!(qpath, QPath::LangItem(LangItem::Range, _))
     && fields.len() == 2
     && fields[0].ident.as_str() == "start"
     && let ExprKind::Lit(ref lit) = fields[0].expr.kind
     && let LitKind::Int(0, LitIntType::Unsuffixed) = lit.node
     && fields[1].ident.as_str() == "end"
     && let ExprKind::Lit(ref lit1) = fields[1].expr.kind
     && let LitKind::Int(10, LitIntType::Unsuffixed) = lit1.node
     && let ExprKind::Block(block, None) = body.kind
     && block.stmts.len() == 1
     && let StmtKind::Semi(e) = block.stmts[0].kind
     && let ExprKind::Break(destination, None) = e.kind
     && let Some(label) = destination.label
     && label.ident.as_str() == "'label"
     && block.expr.is_none()
+    // report your lint here
+}
+}
+if let ExprKind::Loop(body, Some(label), LoopSource::ForLoop, _) = expr.kind
+    && body.stmts.len() == 1
+    && let StmtKind::Expr(e) = body.stmts[0].kind
+    && let ExprKind::Match(scrutinee, arms, MatchSource::ForLoopDesugar) = e.kind
+    && let ExprKind::Call(func, args) = scrutinee.kind
+    && let ExprKind::Path(ref qpath) = func.kind
+    && matches!(qpath, QPath::LangItem(LangItem::IteratorNext, _))
+    && args.len() == 1
+    && let ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mut, inner) = args[0].kind
+    && let ExprKind::Path(ref qpath1) = inner.kind
+    && match_qpath(qpath1, &["iter"])
+    && arms.len() == 2
+    && let PatKind::Struct(ref qpath2, fields, false) = arms[0].pat.kind
+    && matches!(qpath2, QPath::LangItem(LangItem::OptionNone, _))
+    && fields.is_empty()
+    && arms[0].guard.is_none()
+    && let ExprKind::Break(destination, None) = arms[0].body.kind
+    && destination.label.is_none()
+    && let PatKind::Struct(ref qpath3, fields1, false) = arms[1].pat.kind
+    && matches!(qpath3, QPath::LangItem(LangItem::OptionSome, _))
+    && fields1.len() == 1
+    && fields1[0].ident.as_str() == "0"
+    && let PatKind::Wild = fields1[0].pat.kind
+    && arms[1].guard.is_none()
+    && let ExprKind::Block(block, None) = arms[1].body.kind
+    && block.stmts.len() == 1
+    && let StmtKind::Semi(e1) = block.stmts[0].kind
+    && let ExprKind::Break(destination1, None) = e1.kind
+    && let Some(label1) = destination1.label
+    && label1.ident.as_str() == "'label"
+    && block.expr.is_none()
+    && body.expr.is_none()
+    && label.ident.as_str() == "'label"
     // report your lint here
 }
 }
 if let Some(higher::While { condition: condition, body: body }) = higher::While::hir(expr)
     && let ExprKind::Path(ref qpath) = condition.kind
     && match_qpath(qpath, &["a"])
     && let ExprKind::Block(block, None) = body.kind
     && block.stmts.len() == 1
     && let StmtKind::Semi(e) = block.stmts[0].kind
     && let ExprKind::Break(destination, None) = e.kind
     && destination.label.is_none()
     && block.expr.is_none()
     // report your lint here
 }
 }
 if let Some(higher::WhileLet { let_pat: let_pat, let_expr: let_expr, if_then: if_then }) = higher::WhileLet::hir(expr)
     && let PatKind::Lit(lit_expr) = let_pat.kind
     && let ExprKind::Lit(ref lit) = lit_expr.kind
     && let LitKind::Bool(true) = lit.node
     && let ExprKind::Path(ref qpath) = let_expr.kind
     && match_qpath(qpath, &["a"])
     && let ExprKind::Block(block, None) = if_then.kind
     && block.stmts.len() == 1
     && let StmtKind::Semi(e) = block.stmts[0].kind
     && let ExprKind::Break(destination, None) = e.kind
     && destination.label.is_none()
     && block.expr.is_none()
     // report your lint here
 }
 }
 if let ExprKind::Loop(body, None, LoopSource::Loop, _) = expr.kind
     && body.stmts.len() == 1
     && let StmtKind::Semi(e) = body.stmts[0].kind
     && let ExprKind::Break(destination, None) = e.kind
     && destination.label.is_none()
     && body.expr.is_none()
     // report your lint here
 }
 


The actual stdout differed from the expected stdout.
Actual stdout saved to /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/test/ui/author/loop.stage-id.stdout
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args author/loop.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "tests/ui/author/loop.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/test/ui" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/test/ui/author/loop.stage-id" "-A" "unused" "--emit=metadata" "-Dwarnings" "-Zui-testing" "-L" "dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "-L" "dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--extern" "rustc_semver=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/librustc_semver-963bbd3f89834643.rlib" "--extern" "futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-7a940bbcc1286cdc.rlib" "--extern" "syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-7b5cc279bd04eca8.rlib" "--extern" "if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-03f75cdc6d4d3afc.rlib" "--extern" "itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-b6f83e8bf7b1d2e3.rlib" "--extern" "clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-992c1a552e84eb88.rlib" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-e0524b7e2611e851.rlib" "--extern" "serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-edad3f22f7291185.rlib" "--extern" "serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libserde_derive-f7c308c5f8e7b09c.so" "--extern" "regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-619ac20e364f2b2c.rlib" "--extern" "parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f2e351c56701685e.rlib" "--extern" "clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-ec78d46878d61bf1.rlib" "--extern" "quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-e39b631048351f36.rlib" "--extern" "derive_new=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libderive_new-a2f771a227dafeba.so" "--edition=2021" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/test/ui/author/loop.stage-id.aux"
------------------------------------------
------------------------------------------
if let Some(higher::ForLoop { pat: pat, arg: arg, body: body, .. }) = higher::ForLoop::hir(expr)
    && let PatKind::Binding(BindingAnnotation::NONE, _, name, None) = pat.kind
    && name.as_str() == "y"
    && let ExprKind::Struct(qpath, fields, None) = arg.kind
    && matches!(qpath, QPath::LangItem(LangItem::Range, _))
    && fields.len() == 2
    && fields[0].ident.as_str() == "start"
    && let ExprKind::Lit(ref lit) = fields[0].expr.kind
    && let LitKind::Int(0, LitIntType::Unsuffixed) = lit.node
    && fields[1].ident.as_str() == "end"
    && let ExprKind::Lit(ref lit1) = fields[1].expr.kind
    && let LitKind::Int(10, LitIntType::Unsuffixed) = lit1.node
    && let ExprKind::Block(block, None) = body.kind
    && block.stmts.len() == 1
    && let StmtKind::Local(local) = block.stmts[0].kind
    && let Some(init) = local.init
    && let ExprKind::Path(ref qpath1) = init.kind
    && match_qpath(qpath1, &["y"])
    && let PatKind::Binding(BindingAnnotation::NONE, _, name1, None) = local.pat.kind
    && name1.as_str() == "z"
    && block.expr.is_none()
    // report your lint here
}
}
if let ExprKind::Loop(body, None, LoopSource::ForLoop, _) = expr.kind
    && body.stmts.len() == 1
    && let StmtKind::Expr(e) = body.stmts[0].kind
    && let ExprKind::Match(scrutinee, arms, MatchSource::ForLoopDesugar) = e.kind
    && let ExprKind::Call(func, args) = scrutinee.kind
    && let ExprKind::Path(ref qpath) = func.kind
    && matches!(qpath, QPath::LangItem(LangItem::IteratorNext, _))
    && args.len() == 1
    && let ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mut, inner) = args[0].kind
    && let ExprKind::Path(ref qpath1) = inner.kind
    && match_qpath(qpath1, &["iter"])
    && arms.len() == 2
    && let PatKind::Struct(ref qpath2, fields, false) = arms[0].pat.kind
    && matches!(qpath2, QPath::LangItem(LangItem::OptionNone, _))
    && fields.is_empty()
    && arms[0].guard.is_none()
    && let ExprKind::Break(destination, None) = arms[0].body.kind
    && destination.label.is_none()
    && let PatKind::Struct(ref qpath3, fields1, false) = arms[1].pat.kind
    && matches!(qpath3, QPath::LangItem(LangItem::OptionSome, _))
    && fields1.len() == 1
    && fields1[0].ident.as_str() == "0"
    && let PatKind::Binding(BindingAnnotation::NONE, _, name, None) = fields1[0].pat.kind
    && name.as_str() == "y"
    && arms[1].guard.is_none()
    && let ExprKind::Block(block, None) = arms[1].body.kind
    && block.stmts.len() == 1
    && let StmtKind::Local(local) = block.stmts[0].kind
    && let Some(init) = local.init
    && let ExprKind::Path(ref qpath4) = init.kind
    && match_qpath(qpath4, &["y"])
    && let PatKind::Binding(BindingAnnotation::NONE, _, name1, None) = local.pat.kind
    && name1.as_str() == "z"
    && block.expr.is_none()
    && body.expr.is_none()
    // report your lint here
}
}
if let Some(higher::ForLoop { pat: pat, arg: arg, body: body, .. }) = higher::ForLoop::hir(expr)
    && let PatKind::Wild = pat.kind
    && let ExprKind::Struct(qpath, fields, None) = arg.kind
    && matches!(qpath, QPath::LangItem(LangItem::Range, _))
    && fields.len() == 2
    && fields[0].ident.as_str() == "start"
    && let ExprKind::Lit(ref lit) = fields[0].expr.kind
    && let LitKind::Int(0, LitIntType::Unsuffixed) = lit.node
    && fields[1].ident.as_str() == "end"
    && let ExprKind::Lit(ref lit1) = fields[1].expr.kind
    && let LitKind::Int(10, LitIntType::Unsuffixed) = lit1.node
    && let ExprKind::Block(block, None) = body.kind
    && block.stmts.len() == 1
    && let StmtKind::Semi(e) = block.stmts[0].kind
    && let ExprKind::Break(destination, None) = e.kind
    && destination.label.is_none()
    && block.expr.is_none()
    // report your lint here
}
}
if let ExprKind::Loop(body, None, LoopSource::ForLoop, _) = expr.kind
    && body.stmts.len() == 1
    && let StmtKind::Expr(e) = body.stmts[0].kind
    && let ExprKind::Match(scrutinee, arms, MatchSource::ForLoopDesugar) = e.kind
    && let ExprKind::Call(func, args) = scrutinee.kind
    && let ExprKind::Path(ref qpath) = func.kind
    && matches!(qpath, QPath::LangItem(LangItem::IteratorNext, _))
    && args.len() == 1
    && let ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mut, inner) = args[0].kind
    && let ExprKind::Path(ref qpath1) = inner.kind
    && match_qpath(qpath1, &["iter"])
    && arms.len() == 2
    && let PatKind::Struct(ref qpath2, fields, false) = arms[0].pat.kind
    && matches!(qpath2, QPath::LangItem(LangItem::OptionNone, _))
    && fields.is_empty()
    && arms[0].guard.is_none()
    && let ExprKind::Break(destination, None) = arms[0].body.kind
    && destination.label.is_none()
    && let PatKind::Struct(ref qpath3, fields1, false) = arms[1].pat.kind
    && matches!(qpath3, QPath::LangItem(LangItem::OptionSome, _))
    && fields1.len() == 1
    && fields1[0].ident.as_str() == "0"
    && let PatKind::Wild = fields1[0].pat.kind
    && arms[1].guard.is_none()
    && let ExprKind::Block(block, None) = arms[1].body.kind
    && block.stmts.len() == 1
    && let StmtKind::Semi(e1) = block.stmts[0].kind
    && let ExprKind::Break(destination1, None) = e1.kind
    && destination1.label.is_none()
    && block.expr.is_none()
    && body.expr.is_none()
    // report your lint here
}
}
if let Some(higher::ForLoop { pat: pat, arg: arg, body: body, .. }) = higher::ForLoop::hir(expr)
    && let PatKind::Wild = pat.kind
    && let ExprKind::Struct(qpath, fields, None) = arg.kind
    && matches!(qpath, QPath::LangItem(LangItem::Range, _))
    && fields.len() == 2
    && fields[0].ident.as_str() == "start"
    && let ExprKind::Lit(ref lit) = fields[0].expr.kind
    && let LitKind::Int(0, LitIntType::Unsuffixed) = lit.node
    && fields[1].ident.as_str() == "end"
    && let ExprKind::Lit(ref lit1) = fields[1].expr.kind
    && let LitKind::Int(10, LitIntType::Unsuffixed) = lit1.node
    && let ExprKind::Block(block, None) = body.kind
    && block.stmts.len() == 1
    && let StmtKind::Semi(e) = block.stmts[0].kind
    && let ExprKind::Break(destination, None) = e.kind
    && let Some(label) = destination.label
    && label.ident.as_str() == "'label"
    && block.expr.is_none()
    // report your lint here
}
}
if let ExprKind::Loop(body, Some(label), LoopSource::ForLoop, _) = expr.kind
    && body.stmts.len() == 1
    && let StmtKind::Expr(e) = body.stmts[0].kind
    && let ExprKind::Match(scrutinee, arms, MatchSource::ForLoopDesugar) = e.kind
    && let ExprKind::Call(func, args) = scrutinee.kind
    && let ExprKind::Path(ref qpath) = func.kind
    && matches!(qpath, QPath::LangItem(LangItem::IteratorNext, _))
    && args.len() == 1
    && let ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mut, inner) = args[0].kind
    && let ExprKind::Path(ref qpath1) = inner.kind
    && match_qpath(qpath1, &["iter"])
    && arms.len() == 2
    && let PatKind::Struct(ref qpath2, fields, false) = arms[0].pat.kind
    && matches!(qpath2, QPath::LangItem(LangItem::OptionNone, _))
    && fields.is_empty()
    && arms[0].guard.is_none()
    && let ExprKind::Break(destination, None) = arms[0].body.kind
    && destination.label.is_none()
    && let PatKind::Struct(ref qpath3, fields1, false) = arms[1].pat.kind
    && matches!(qpath3, QPath::LangItem(LangItem::OptionSome, _))
    && fields1.len() == 1
    && fields1[0].ident.as_str() == "0"
    && let PatKind::Wild = fields1[0].pat.kind
    && arms[1].guard.is_none()
    && let ExprKind::Block(block, None) = arms[1].body.kind
    && block.stmts.len() == 1
    && let StmtKind::Semi(e1) = block.stmts[0].kind
    && let ExprKind::Break(destination1, None) = e1.kind
    && let Some(label1) = destination1.label
    && label1.ident.as_str() == "'label"
    && block.expr.is_none()
    && body.expr.is_none()
    && label.ident.as_str() == "'label"
    // report your lint here
}
}
if let Some(higher::While { condition: condition, body: body }) = higher::While::hir(expr)
    && let ExprKind::Path(ref qpath) = condition.kind
    && match_qpath(qpath, &["a"])
    && let ExprKind::Block(block, None) = body.kind
    && block.stmts.len() == 1
    && let StmtKind::Semi(e) = block.stmts[0].kind
    && let ExprKind::Break(destination, None) = e.kind
    && destination.label.is_none()
    && block.expr.is_none()
    // report your lint here
}
}
if let Some(higher::WhileLet { let_pat: let_pat, let_expr: let_expr, if_then: if_then }) = higher::WhileLet::hir(expr)
    && let PatKind::Lit(lit_expr) = let_pat.kind
    && let ExprKind::Lit(ref lit) = lit_expr.kind
    && let LitKind::Bool(true) = lit.node
    && let ExprKind::Path(ref qpath) = let_expr.kind
    && match_qpath(qpath, &["a"])
    && let ExprKind::Block(block, None) = if_then.kind
    && block.stmts.len() == 1
    && let StmtKind::Semi(e) = block.stmts[0].kind
    && let ExprKind::Break(destination, None) = e.kind
    && destination.label.is_none()
    && block.expr.is_none()
    // report your lint here
}
}
if let ExprKind::Loop(body, None, LoopSource::Loop, _) = expr.kind
    && body.stmts.len() == 1
    && let StmtKind::Semi(e) = body.stmts[0].kind
    && let ExprKind::Break(destination, None) = e.kind
    && destination.label.is_none()
    && body.expr.is_none()
    // report your lint here
}

------------------------------------------
```

</details>
